fairlearn/postprocessing/_cvxpy_threshold_optimizer.py

```diff
@@ -0,0 +1,395 @@
+"""Threshold optimizer with relaxed fairness constraints.
+
+TODO
+----
+- Add option for constraining only equality of FPR or TPR (currently it must be 
+both -> equal odds);
+- Add option for constraining equality of positive predictions (independence
+criterion, aka demographic parity);
+- Add option to use l1 or linf distances for maximum tolerance between points.
+  - Currently 'equal_odds' is defined using l-infinity distance (max between
+  TPR and FPR distances);
+
+"""
+import logging
+from itertools import product
+from typing import Callable
+
+import numpy as np
+from sklearn.metrics import roc_curve
+from sklearn.base import BaseEstimator, MetaEstimatorMixin
+
+from fairlearn.utils._input_validation import _validate_and_reformat_input
+from fairlearn.reductions._moments.error_rate import _MESSAGE_BAD_COSTS
+
+from ._cvxpy_utils import compute_equal_odds_optimum
+from ._roc_utils import (
+    roc_convex_hull,
+    calc_cost_of_point,
+)
+from ._randomized_classifiers import (  # TODO: try to use fairlearn's InterpolatedThreshold instead of our classifier API
+    RandomizedClassifier,
+    EnsembleGroupwiseClassifiers,
+)
+
+
+class _RelaxedThresholdOptimizer(BaseEstimator, MetaEstimatorMixin):
+    r"""Class to encapsulate all the logic needed to compute the optimal
+    postprocessing to fulfill fairness constraints with some optional
+    tolerance.
+
+    The method amounts to finding the set of (potentially randomized) 
+    group-specific decision thresholds that maximize some goal (e.g., accuracy),
+    given a maximum tolerance (or slack) on the fairness constraint fulfillment.
+    
+    This optimization problem amounts to a Linear Program (LP) as detailed in 
+    [1]_. Solving the LP requires installing `cvxpy`.
+
+    Read more in the :ref:`User Guide <postprocessing>`.
+
+    Parameters
+    ----------
+    estimator : object
+        A `scikit-learn compatible estimator <https://scikit-learn.org/stable/developers/develop.html#estimators>`_  # noqa
+        whose output will be postprocessed.
+        The estimator should output real-valued scores, as postprocessing 
+        results will be extremely poor when performed over binarized 
+        predictions.
+
+    tolerance : float
+        The absolute tolerance for the equal odds fairness constraint.
+        Will allow for `tolerance` difference between group-wise ROC points.
+
+    objective_costs : dict
+        A dictionary detailing the cost for false positives and false negatives,
+        of the form :code:`{'fp': <fp_cost>, 'fn': <fn_cost>}`. Will use the 0-1
+        loss by default (maximum accuracy).
+
+    grid_size : int, optional
+        The maximum number of ticks (points) in each group's ROC curve, by
+        default 1000. This corresponds to the maximum number of different 
+        thresholds to use over a predictor.
+
+    seed : int
```

Comment:
Changed to `random_state`.

Right now it's used in the `fit` method, not in the predictions.
For example, it's used to create random number generators in each randomized classifier as they are being constructed. This could be done lazily when predict is called, but I think the `RandomizedClassifier` might just disappear during this PR so it's probably not worth it for now.

Reply:
Sorry, didn't realize it was used in `fit`. I had somehow thought only `predict` uses it. Which brings me to another question: Doesn't predict also use it?

Reply:
`predict` also uses it, in the sense that the `fit` method created a randomized thresholded classifier (with some `random_state`) that is then used in `predict`.

I think this implementation had reproducibility in mind, but right now we may want to move the `random_state` to `predict`. I'll leave this thread open.